docs/src/design/decoder/constraints.md

```diff
@@ -108,14 +110,14 @@ In constructing value of $u$ for decoder AIR constraints, we will use the follow
 To simplify constraint description, we define the following variables:
 
 $$
-h_{init4} = \alpha_0 + \alpha_1 \cdot m_{bp} + \alpha_2 \cdot a' + \sum_{i=0}^3(\alpha_{i + 8} \cdot h_i)
+h_{init4} = \alpha_0 + \alpha_1 \cdot m_{bp} + \alpha_2 \cdot a + f_{ctrlb} \cdot \alpha_5 \cdot \sum_{b=0}^6(b_i \cdot 2^i) + \sum_{i=0}^3(\alpha_{i + 8} \cdot h_i)
```

Comment:
Yes this was intentional, I believe there was a mistake here previously.  When we initialise the hasher we should mark the address as `a`.  Then the result is specified at `a + 7` and in the case of linear hashing we absorb again from the rate at `a' = a+8`.  If you look just below this you will be able to see this explained.

Reply:
I'll need to look into this in more detail - but it does seem to me that `a'` was correct. For example, based on the diagram [here](https://0xpolygonmiden.github.io/miden-vm/design/decoder/main.html#join-operation), the address of the block comes from the next row (which is `a'`). The `+7` part indicates the number of rows in the hasher (not in the decoder).

Reply:
yeah I think there is inconsistent use of language here because `a'` is defined as $a' = a + 8$.

Reply:
I've removed the $f_{ctrlb}$ flag and embedded the $\alpha_5 * opcode$ value directly in the $u_{join}$, $u_{split}$, ... terms.  I believe this will yield a lower degree.  I think the previous pattern of having the opcode calculation within $h_{init}$ wouldn't have worked as it would have yielded a higher degree for the $u_{join}$, ... terms.

Reply:
I don't think the language use is inconsistent, but I think you're right that it's a bit confusing. Maybe we should add some clarifying language to emphasize that we expect $a' = a + 8$ during absorption (i.e. "In the above" doesn't mean all of the above), and to remind people that `a` and `a'` refer to the address value in the decoder, which is set based on the referenced hasher rows, which in this cases are 8 rows apart.

Reply:
Yeah I think this confusion was largely driven by my lack of understanding of how control block hashing works in the decoder but maybe some clarification in the text would help.

Reply:
> Yeah I think this confusion was largely driven by my lack of understanding of how control block hashing works in the decoder

I think this indicates a problem with the docs, which I think we might as well address now, unless you guys disagree.

Maybe something like:

"In the above computation of the absorption label $h_{abp}$, we assume that $a' = a + 8$, since $a$ is set using the address of the corresponding hasher row, which in this case should be the row where the absorption begins. This relationship is enforced via separate constraints."

Reply:
Yeah I think an update would be useful.  Is there any reason we don't just explicitly specify $a + 8$ in the equation such that we have:
$h_{abp} = \alpha_0 + \alpha_1 \cdot m_{abp} + \alpha_2 \cdot (a + 8) + \sum_{i=0}^7(\alpha_{i + 8} \cdot h_i)$
I think this is clearer as there is no reuse of the $a'$ term.  We use an offset from a for $h_{res}$, see here:
$h_{res} = \alpha_0 + \alpha_1 \cdot m_{hout} + \alpha_2 \cdot (a + 7) + \sum_{i=0}^3(\alpha_{i + 8} \cdot h_i)$

Reply:
> "In the above computation of the absorption label $h_{abp}$, we assume that $a′=a+8$, since a is set using the address of the corresponding hasher row, which in this case should be the row where the absorption begins. This relationship is enforced via separate constraints."

So I did a little more reading on this. I believe $a$ represents the hasher address of the previous absorption (or initialization).  As such, $a`$ is set to $a + 8$ which is the hasher address for the current absorption. Maybe we could word this as. 

I think we word it as follows:
"In the above computation of the absorption label $h_{abp}$, it should be noted that $a$ represents the hasher address of the previous hasher absorption (or initialization). As such, $a'$ is set to $a + 8$ which is the hasher address for the current absorption. This relationship is enforced via separate constraints."

Reply:
Just to be clear - `a` refers to a column in the decoder, as depicted. The addresses in this column are set by using the address from the hasher chiplet for the corresponding hash initialization/absorption/return. So in this case, $a$, the value of the address in the current row of the decoder, is set to equal the value of the address of the row in the hasher chiplet where the previous absorption (or initialization) occurred, and $a'$ is set to equal the address in the hasher chiplet where the absorption referred to by the $h_{abp}$ label is happening.

I think your rephrase is fine (aside from the nit I left inline at that point), but just wanted to make sure that the above is clear

Reply:
I think the way you've phrased it is more explicit and clearer. I'll update it.

Reply:
I've updated with the phrasing you suggested and also modified the description of $h_{res}$ accordingly to be:

In the above, $a$ represents the address value in the decoder which corresponds to the hasher chiplet address at which the hasher was initialized (or the last absorption took place).  As such, $a + 7$ corresponds to the hasher chiplet address at which the result is returned.